includes/functions.php

```diff
@@ -0,0 +1,293 @@
+<?php
+/**
+ * Common plugin functions
+ *
+ * @package WP_CLI_Cron_Control_Offload
+ */
+
+namespace Automattic\WP\WP_CLI_Cron_Control_Offload;
+use WP_Error;
+
+/**
+ * Create cron event for a given WP-CLI command
+ *
+ * @param string $command WP-CLI command to schedule.
+ * @param int    $timestamp Optional. Unix timestamp to schedule command to run at.
+ * @return int|WP_Error
+ */
+function schedule_cli_command( $command, $timestamp = null ) {
+	$command = validate_command( $command );
+
+	if ( is_wp_error( $command ) ) {
+		return $command;
+	}
+
+	if ( ! is_int( $timestamp ) ) {
+		$timestamp = strtotime( '+30 seconds' );
+	}
+
+	if ( $timestamp <= time() ) {
+		return new WP_Error( 'invalid-timestamp', __( 'Timestamp is in the past.', 'wp-cli-cron-control-offload' ) );
+	}
+
+	$event_args = array(
+		'command' => $command,
+	);
+
+	$scheduled = wp_schedule_single_event( $timestamp, ACTION, $event_args );
+
+	if ( false === $scheduled ) {
+		return new WP_Error( 'not-scheduled', __( 'Command may already be scheduled, or it was blocked via the `schedule_event` filter.', 'wp-cli-cron-control-offload' ) );
+	}
+
+	return $timestamp;
+}
+
+/**
+ * Validate WP-CLI command to be scheduled
+ *
+ * @param string $command WP-CLI command to validate.
+ * @return string|WP_Error
+ */
+function validate_command( $command ) {
+	$command = trim( $command );
+	$command = parse_command( $command );
```

Comment:
I think the parse / implode feel a bit dangerous and maybe even unnecessary. Looking through, there are some things we want to do:

1) find out which command is being called (so we can check against the whitelist/blacklist).
2) prevent bad positionals (`&`, '|', etc.)

I think we could do both those things without the full complexity of breaking out the command and then rebuilding it.

The other option would be to change how the command is executed so that instead of accepting a freeform string, we ask for a explicit structure (i.e. the command needs to be formatted as `{command} {subcommand} {flags}`; anything else fails validation).

Reply:
Most of the logic comes from WP-CLI, and arose from wanting to accomodate `WP_CLI::runcommand()`. It doesn't expect the leading `wp` (trivial), but also passes along many of the global associative arguments, so there's a bit of need to clean the command. WP-CLI also doesn't expose its validation methods for reuse. 

I like your idea to change how the command is accepted. That'll remove a lot of the magic.

Reply:
I'm leaning towards breaking `command`, `subcommand` and `flags` into their own associative arguments, removing all of the magic from parsing the command.